pkg/private/zip/build_zip.py

```diff
@@ -68,53 +69,144 @@ def parse_date(ts):
+    self.zip_file = None
+
+  def make_zipinfo(self, path: str, mode: int):
+    """Create a Zipinfo.
+
+    Args:
+      path: file path
+      mode: file mode
+    """
+    entry_info = zipfile.ZipInfo(filename=path, date_time=self.time_stamp)
+    # See http://www.pkware.com/documents/casestudies/APPNOTE.TXT
+    # denotes UTF-8 encoded file name.
+    entry_info.flag_bits |= 0x800
+
+    # See: https://trac.edgewall.org/attachment/ticket/8919/ZipDownload.patch
+    # external_attr is 4 bytes in size. The high order two bytes represent UNIX
+    # permission and file type bits, while the low order two contain MS-DOS FAT
+    # file attributes.
+    if mode:
+      f_mode = int(mode, 8)
+    else:
+      f_mode = self.default_mode
+    entry_info.external_attr = f_mode << 16
+    return entry_info
+
+  def add_manifest_entry(self, options, entry):
+    """Add an entry to the zip file.
+
+    Args:
+      options: parsed options
+      zip_file: ZipFile to write to
+      entry: manifest entry
+    """
+
+    entry_type, dest, src, mode, user, group = entry
+
+    # Use the pkg_tar mode/owner remaping as a fallback
+    non_abs_path = dest.strip('/')
+    dst_path = _combine_paths(options.directory, non_abs_path)
+    if entry_type == manifest.ENTRY_IS_DIR and not dst_path.endswith('/'):
+      dst_path += '/'
+    entry_info = self.make_zipinfo(path=dst_path, mode=mode)
+
+    if entry_type == manifest.ENTRY_IS_FILE:
+      entry_info.compress_type = zipfile.ZIP_DEFLATED
+      with open(src, 'rb') as src:
+        self.zip_file.writestr(entry_info, src.read())
+    elif entry_type == manifest.ENTRY_IS_DIR:
+      entry_info.compress_type = zipfile.ZIP_STORED
+      # Set directory bits
+      entry_info.external_attr |= (UNIX_DIR_BIT << 16) | MSDOS_DIR_BIT
+      self.zip_file.writestr(entry_info, '')
+    elif entry_type == manifest.ENTRY_IS_LINK:
+      entry_info.compress_type = zipfile.ZIP_STORED
+      # Set directory bits
+      entry_info.external_attr |= (UNIX_SYMLINK_BIT << 16)
+      self.zip_file.writestr(entry_info, src)
+    elif entry_type == manifest.ENTRY_IS_TREE:
+      self.add_tree(src, dst_path, mode)
+    elif entry.entry_type == manifest.ENTRY_IS_EMPTY_FILE:
+      entry_info.compress_type = zipfile.ZIP_DEFLATED
+      self.zip_file.writestr(entry_info, '')
+    else:
+      raise Exception('Unknown type for manifest entry:', entry)
+
+  def add_tree(self, tree_top: str, destpath: str, mode: int):
+    """Add a tree artifact to the zip file.
+
+    Args:
+      tree_top: the top of the tree to add
+      destpath: the path under which to place the files
+      mode: if not None, file mode to apply to all files
+    """
+
+    # We expect /-style paths.
+    tree_top = os.path.normpath(tree_top).replace(os.path.sep, '/')
+
+    # Again, we expect /-style paths.
+    dest = destpath.strip('/')  # redundant, dests should never have / here
+    dest = os.path.normpath(dest).replace(os.path.sep, '/')
```

Comment:
Can paths end with `\` on Windows platforms here?  Maybe swap the order of these functions?

Reply:
destpath would be created by our tools, so it will have a '/'.
Then normpath may turn / into \, so we have to turn it back. 
I thinks this works right as is.